### PR TITLE
Experimental instrumentation for the OpenTelemetry Tracing API

### PR DESCRIPTION
**Pull request by tylerbenson**
This can be used as an alternative to OpenTracing for custom instrumentation.

The integration works by hijacking the `OpenTelemetry.getTracerProvider` and `OpenTelemetry.getPropagators` method return values and replacing it with our own.

To enable:
* System Property: `-Ddd.integration.opentelemetry-beta.enabled=true`
* Environment Variable: `DD_INTEGRATION_OPENTELEMETRY_BETA_ENABLED=true`

The actual mapping semantics are subject to change as the API matures.

Events are completely ignored as there's no analogous feature in Datadog APM. This also won't impact the existing MeterProvider.

### CODE REVIEW

File: dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelContextPropagators.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+import io.opentelemetry.context.propagation.ContextPropagators;
+import io.opentelemetry.context.propagation.HttpTextFormat;
+import io.opentelemetry.trace.DefaultSpan;
+import io.opentelemetry.trace.Span;
+import io.opentelemetry.trace.TracingContextUtils;
+import java.util.Arrays;
+import java.util.List;
+
+public class OtelContextPropagators implements ContextPropagators {
+  public static final OtelContextPropagators INSTANCE = new OtelContextPropagators();
```
Comment (devinsba) [inline dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelContextPropagators.java:16]: A matching private constructor would be good

File: dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelSpan.java [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+  @Override
+  public void addEvent(final Event event) {}
+
+  @Override
+  public void addEvent(final Event event, final long timestamp) {}
+
+  @Override
+  public void setStatus(final Status status) {
+    if (!status.isOk()) {
+      delegate.setError(true);
```
Comment (devinsba) [inline dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelSpan.java:86]: Doesn't this depend on the type of span? IE: a client sees 400 as an error where a server doesn't

File: dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelTracerProvider.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  private final TypeConverter converter = new TypeConverter();
+
+  @Override
+  public Tracer get(final String instrumentationName) {
+    return get(instrumentationName, null);
+  }
+
+  @Override
+  public Tracer get(final String instrumentationName, final String instrumentationVersion) {
+    return new OtelTracer(instrumentationName, AgentTracer.get(), converter);
```
Comment (devinsba) [inline dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelTracerProvider.java:19]: We should probably cache this instead of returning a new one everytime to protect from bad instrumentation

File: dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OpenTelemetryInstrumentation.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        named("getTracerProvider").and(returns(named("io.opentelemetry.trace.TracerProvider"))),
+        OpenTelemetryInstrumentation.class.getName() + "$TracerProviderAdvice");
+    transformers.put(
+        named("getPropagators")
+            .and(returns(named("io.opentelemetry.context.propagation.ContextPropagators"))),
+        OpenTelemetryInstrumentation.class.getName() + "$ContextPropagatorsAdvice");
+    return transformers;
+  }
+
+  public static class TracerProviderAdvice {
```
Comment (richardstartin) [inline dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OpenTelemetryInstrumentation.java:69]: Could we move away from having lots of small nested classes with long names?

File: dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelContextPropagators.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+    private final HttpTextFormat.Getter<C> getter;
+
+    private OtelGetter(final HttpTextFormat.Getter<C> getter) {
+      this.getter = getter;
+    }
+
+    @Override
+    public Iterable<String> keys(final C carrier) {
+      // TODO: Otel doesn't expose the keys, so we have to rely on hard coded keys.
```
Comment (richardstartin) [inline dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelContextPropagators.java:99]: The thing about iterating over the keys and then checking for each possible thing that we care about it is that it's quadratic time and works against pretty much every HTTP API which presents headers to you as a HashMap. Maybe it makes sense not to encourage users to consume headers the way we do?

File: dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelContextPropagators.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+      // https://github.com/open-telemetry/opentelemetry-specification/issues/433
+      return KEYS;
+    }
+
+    @Override
+    public String get(final C carrier, final String key) {
+      return getter.get(carrier, key);
```
Comment (richardstartin) [inline dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelContextPropagators.java:106]: Perhaps we should be aiming to use this more since every HTTP API can do this quickly.

File: dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelScope.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
+package datadog.trace.instrumentation.opentelemetry;
+
+import datadog.trace.bootstrap.instrumentation.api.AgentScope;
+import datadog.trace.context.TraceScope;
+import io.opentelemetry.context.Scope;
+
+public class OtelScope implements Scope, TraceScope {
+  private final AgentScope delegate;
```
Comment (richardstartin) [inline dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelScope.java:8]: Scopes all the way down?

File: dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelSpan.java [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+
+  @Override
+  public void end() {
+    delegate.finish();
+  }
+
+  @Override
+  public void end(final EndSpanOptions endOptions) {
+    if (endOptions.getEndTimestamp() > 0) {
+      delegate.finish(TimeUnit.NANOSECONDS.toMicros(endOptions.getEndTimestamp()));
```
Comment (richardstartin) [inline dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelSpan.java:104]: Which timezone? Is there a guarantee it's UTC?

File: dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/TypeConverter.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.trace.Span;
+import io.opentelemetry.trace.SpanContext;
+
+// Centralized place to do conversions
+public class TypeConverter {
+  // TODO maybe add caching to reduce new objects being created
+
+  public AgentSpan toAgentSpan(final Span span) {
+    if (span == null) {
```
Comment (richardstartin) [inline dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/TypeConverter.java:15]:
```java
    if (span instanceof OtelSpan) {
      return ((OtelSpan) span).getDelegate();
    }
    return null == span ? null : AgentTracer.NoopAgentSpan.INSTANCE;
```

File: dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/TypeConverter.java [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+  public SpanContext toSpanContext(final AgentSpan.Context context) {
+    if (context == null) {
+      return null;
+    }
+    return new OtelSpanContext(context);
+  }
+
+  public AgentSpan.Context toContext(final SpanContext spanContext) {
+    if (spanContext == null) {
```
Comment (richardstartin) [inline dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/TypeConverter.java:47]:
```java
    if (spanContext instanceof OtelSpanContext) {
      return ((OtelSpanContext) spanContext).getDelegate();
    } 
    return null == spanContext ? null : AgentTracer.NoopContext.INSTANCE;
```

File: dd-java-agent/instrumentation/opentelemetry/src/test/groovy/OpenTelemetryTest.groovy [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+import datadog.trace.core.propagation.ExtractedContext
+import io.grpc.Context
+import io.opentelemetry.OpenTelemetry
+import io.opentelemetry.context.propagation.HttpTextFormat
+import io.opentelemetry.trace.Status
+import io.opentelemetry.trace.TracingContextUtils
+import spock.lang.Subject
+
+class OpenTelemetryTest extends AgentTestRunner {
+  static {
```
Comment (richardstartin) [inline dd-java-agent/instrumentation/opentelemetry/src/test/groovy/OpenTelemetryTest.groovy:16]: Since this can leak to other tests, is it a time bomb? Do we need a much easier way of controlling environmental conditions?

File: dd-java-agent/instrumentation/opentelemetry/src/test/groovy/OpenTelemetryTest.groovy [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+  }
+
+  @Subject
+  def tracer = OpenTelemetry.tracerProvider.get("test-inst")
+  def httpPropagator = OpenTelemetry.getPropagators().httpTextFormat
+
+  def "test span"() {
+    setup:
+    def builder = tracer.spanBuilder("some name")
+    if (tagBuilder) {
```
Comment (richardstartin) [inline dd-java-agent/instrumentation/opentelemetry/src/test/groovy/OpenTelemetryTest.groovy:27]: I'm not sure how I feel about this much logic in tests. I haven't encountered it in other code bases.

File: dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelTracer.java [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+  public Span.Builder spanBuilder(final String spanName) {
+    return new SpanBuilder(spanName);
+  }
+
+  private class SpanBuilder implements Span.Builder {
+    private final AgentTracer.SpanBuilder delegate;
+    private boolean parentSet = false;
+
+    public SpanBuilder(final String spanName) {
+      delegate = tracer.buildSpan(tracerName).withResourceName(spanName);
```
Comment (richardstartin) [inline dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelTracer.java:50]: how deep can this chain go before a field gets set?

File: dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelSpan.java (hunk 2 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelSpan.java:86]: This isn't the regular http status codes... this is an OTel specific Status (more similar to grpc) https://github.com/open-telemetry/opentelemetry-java/blob/master/api/src/main/java/io/opentelemetry/trace/Status.java

File: dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelContextPropagators.java (hunk 5 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelContextPropagators.java:99]: I figured you'd appreciate this... the problem is it doesn't work with baggage which relies on a prefix.

File: dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelTracerProvider.java (hunk 3 above)
Comment (richardstartin) [reply dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelTracerProvider.java:19]: There's a nice very low overhead cache on #1603, maybe we can use that once that's merged, but merge this without implementing caching yet?

File: dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OpenTelemetryInstrumentation.java (hunk 4 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OpenTelemetryInstrumentation.java:69]: This should be a separate discussion and done in a consistent fashion.  Meanwhile, this is consistent with the rest of the repo style.

File: dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelScope.java (hunk 7 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelScope.java:8]: It's a wrapper/bridge class... similar to what is done in the opentracing instrumentation.

File: dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelSpan.java (hunk 8 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelSpan.java:104]: I'm actually not sure... I think it fully depends on what the user provides which we can't control.

File: dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelContextPropagators.java (hunk 5 above)
Comment (richardstartin) [reply dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelContextPropagators.java:99]: Then that's a design flaw in the codec

File: dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelSpan.java (hunk 8 above)
Comment (richardstartin) [reply dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelSpan.java:104]: It doesn't make sense to compare timestamps taken relative to different clocks, so whoever computes durations needs to be in control of the clocks.

File: dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelContextPropagators.java (hunk 1 above)
Comment (dougqh) [reply dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelContextPropagators.java:16]: Yes, and make this final as well.

File: dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelTracerProvider.java (hunk 3 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelTracerProvider.java:22]: I added a todo.

File: dd-java-agent/instrumentation/opentelemetry/src/test/groovy/OpenTelemetryTest.groovy (hunk 11 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/opentelemetry/src/test/groovy/OpenTelemetryTest.groovy:16]: I tried clearing this and other similar tests out in cleanup, but it caused other test problems, likely because of a race condition with other tests.  Not sure though.  Either way, I don't want to deal with this problem here.

File: dd-java-agent/instrumentation/opentelemetry/src/test/groovy/OpenTelemetryTest.groovy (hunk 12 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/opentelemetry/src/test/groovy/OpenTelemetryTest.groovy:27]: Does this look better now?

**Comment by tylerbenson** [comment]
I think I've addressed most of the feedback above.  If you think I've missed something that you'd like to see in this PR, please chime in again.

**Comment by tylerbenson** [comment]
I'm not opposed to the idea, but maybe when we are ready to remove the experimental/beta part and want to target specific versions...

File: dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelScope.java [hunk 14]
```diff
@@ -0,0 +1,37 @@
+package datadog.trace.instrumentation.opentelemetry;
+
+import datadog.trace.bootstrap.instrumentation.api.AgentScope;
+import datadog.trace.context.TraceScope;
+import io.opentelemetry.context.Scope;
+
+public class OtelScope implements Scope, TraceScope {
```
Comment (dougqh) [inline dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelScope.java:7]: I wouldn't expect our Otel objects to implement our objects -- just the Otel objects.
I'm curious why we're implementing both interfaces.

File: dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelScope.java [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+  public Continuation capture() {
+    return delegate.capture();
+  }
+
+  @Override
+  public void close() {
+    delegate.close();
+  }
+
+  @Override
```
Comment (dougqh) [inline dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelScope.java:24]: I'm surprised to see the async propagation methods on this class.  I realize it follows from implementing datadog's TraceScope, but wonder why we need to do that.

File: dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelSpan.java [hunk 16]
```diff
@@ -0,0 +1,223 @@
...
+        delegate.setTag(key, value.getDoubleValue());
+        break;
+      case STRING:
+        delegate.setTag(key, value.getStringValue());
+        break;
+      case BOOLEAN:
+        delegate.setTag(key, value.getBooleanValue());
+        break;
+      default:
+        // Unsupported.... Ignoring.
```
Comment (dougqh) [inline dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelSpan.java:60]: I wonder if the unsupported methods / cases should an option to raise UnsupportedOperationException.
Not raising an exception bit us in another situation recently.

File: dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelSpan.java [hunk 17]
```diff
@@ -0,0 +1,223 @@
...
+import io.opentelemetry.common.AttributeValue;
+import io.opentelemetry.trace.EndSpanOptions;
+import io.opentelemetry.trace.Event;
+import io.opentelemetry.trace.Span;
+import io.opentelemetry.trace.SpanContext;
+import io.opentelemetry.trace.Status;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+public class OtelSpan implements Span, MutableSpan {
```
Comment (dougqh) [inline dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelSpan.java:15]: As with Scope, I wouldn't expect this to implement MutableSpan.

File: dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelSpan.java [hunk 18]
```diff
@@ -0,0 +1,223 @@
...
+  }
+
+  @Override
+  public String getOperationName() {
+    return delegate.getOperationName();
+  }
+
+  @Override
+  public MutableSpan setOperationName(final String serviceName) {
+    return delegate.setOperationName(serviceName);
```
Comment (dougqh) [inline dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelSpan.java:141]: This looks wrong to me.
I would expect this to return "this" -- not unwrap delegate.

File: dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelSpanContext.java [hunk 19]
```diff
@@ -0,0 +1,47 @@
+package datadog.trace.instrumentation.opentelemetry;
+
+import datadog.trace.bootstrap.instrumentation.api.AgentSpan;
+import io.opentelemetry.trace.SpanContext;
+import io.opentelemetry.trace.SpanId;
+import io.opentelemetry.trace.TraceFlags;
+import io.opentelemetry.trace.TraceId;
+import io.opentelemetry.trace.TraceState;
+
+public class OtelSpanContext extends SpanContext {
```
Comment (dougqh) [inline dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelSpanContext.java:10]: This doesn't implement AgentSpan.Context?  I generally don't think we should implement our interfaces here, but the inconsistency confuses me.

File: dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelTracer.java [hunk 20]
```diff
@@ -0,0 +1,163 @@
...
+import java.util.concurrent.TimeUnit;
+
+public class OtelTracer implements Tracer {
+  private final String tracerName;
+  private final AgentTracer.TracerAPI tracer;
+  private final TypeConverter converter;
+
+  OtelTracer(
+      final String tracerName, final AgentTracer.TracerAPI tracer, final TypeConverter converter) {
+    this.tracerName = tracerName;
```
Comment (dougqh) [inline dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/OtelTracer.java:23]: I'd like if we could push the information down into our tracers, too, but that can be another PR.

File: dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/TypeConverter.java [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+
+import datadog.trace.bootstrap.instrumentation.api.AgentScope;
+import datadog.trace.bootstrap.instrumentation.api.AgentSpan;
+import datadog.trace.bootstrap.instrumentation.api.AgentTracer;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.trace.Span;
+import io.opentelemetry.trace.SpanContext;
+
+// Centralized place to do conversions
+public class TypeConverter {
```
Comment (dougqh) [inline dd-java-agent/instrumentation/opentelemetry/src/main/java/datadog/trace/instrumentation/opentelemetry/TypeConverter.java:11]: I feel like this might be making things more complicated than they need to be.

I don't see much value in centralizing this logic given that the conversion to / from should be a 1-liner.
I'm also worried by the addition of more instanceof checks which we should avoid, but also just seem unnecessary in this case.
